test/test_autograd.py

```diff
@@ -6188,6 +6192,258 @@ def foo(a):
+            # not always their tangents.
+
+            # Check that inplace ops do not update non-differentiable views
+            cloned_foo = foo.clone()
+            cloned_bar = bar.clone()
+            # Non differentiable view
+            dual = fwAD.make_dual(cloned_foo, cloned_bar)
+            dual *= 2
+            # Check that non differentiable view was not updated
+            self.assertIsNone(fwAD.unpack_dual(foo)[1])
+            # Check that the computed result is correct
+            self.assertEqual(cloned_bar, bar * 2)
+            self.assertEqual(fwAD.unpack_dual(dual)[1], bar * 2)
+            self.assertEqual(cloned_foo, foo * 2)
+            self.assertEqual(fwAD.unpack_dual(dual)[0], foo * 2)
+            # Other non differentiable view
+            dual_primal, dual_tangent = fwAD.unpack_dual(dual)
+            self.assertIsNone(fwAD.unpack_dual(dual_primal)[1])
+            self.assertIsNone(fwAD.unpack_dual(dual_tangent)[1])
+            dual_primal *= 2
+            # Ensure dual's tangent did not change
+            self.assertEqual(fwAD.unpack_dual(dual)[0], foo * 4)
+            self.assertEqual(fwAD.unpack_dual(dual)[1], bar * 2)
+            dual_tangent *= 2
+            # Ensure dual's primal did not change
+            self.assertEqual(fwAD.unpack_dual(dual)[0], foo * 4)
+            self.assertEqual(fwAD.unpack_dual(dual)[1], bar * 4)
+
+            # Check that inplace ops do update differentiable view but stop at non differentiable ones
+            cloned_foo = foo.clone()
+            cloned_bar = bar.clone()
+            # A non differentiable view
+            dual = fwAD.make_dual(cloned_foo, cloned_bar)
+            # A differentiable view
+            view = dual.narrow(0, 0, 1)
+            view *= 2
+            # Check that non differentiable view was not updated
+            self.assertIsNone(fwAD.unpack_dual(foo)[1])
+            # Check that differentiable view was updated
+            self.assertEqual(fwAD.unpack_dual(dual)[1], torch.tensor([2., 1.]))
+            self.assertEqual(fwAD.unpack_dual(view)[1], torch.tensor([2.]))
+
+            # Check that we track differentiable view even for Tensors that are not dual
+            # Basic inplace should update baz
+            baz = torch.rand(2)
+            baz += dual
+            self.assertEqual(fwAD.unpack_dual(baz)[1], fwAD.unpack_dual(dual)[1])
+            # Updates on view should as well
+            baz = torch.rand(2)
+            baz[0] = dual[0]
+            self.assertEqual(fwAD.unpack_dual(baz)[1][0], fwAD.unpack_dual(dual)[1][0])
+            # Unused values get a gradient of 0
+            self.assertEqual(fwAD.unpack_dual(baz)[1][1], 0.)
+
+            # Check that backward non-differentiable views don't prevent gradient update
+            baz = torch.rand(2)
+            view = baz.detach()
+            view += dual
+            self.assertEqual(fwAD.unpack_dual(baz)[1], fwAD.unpack_dual(dual)[1])
```

Comment:
It seems to me like a better testing strategy would be to have a small Python-only implementation of forward mode AD, and then just demonstrate that the C++ and Python version's behaviors coincide exactly.  You have to do a lot of manual verification of results that wouldn't be necessary if you had a reference implementation to compare against.